servant-server/src/Servant/Server/Internal/RoutingApplication.hs

```diff
@@ -155,32 +159,111 @@ data Delayed :: * -> * where
   Delayed :: IO (RouteResult a)
           -> IO (RouteResult ())
           -> IO (RouteResult b)
-          -> (a -> b -> RouteResult c)
-          -> Delayed c
+          -> IO (RouteResult c)
+          -> (a -> b -> c -> RouteResult d)
+          -> Delayed d
 
 instance Functor Delayed where
-   fmap f (Delayed a b c g) = Delayed a b c ((fmap.fmap.fmap) f g)
+   fmap f (Delayed a b c d g) = Delayed a b c d ((fmap.fmap.fmap.fmap) f g)
 
 -- | Add a capture to the end of the capture block.
 addCapture :: Delayed (a -> b)
            -> IO (RouteResult a)
            -> Delayed b
-addCapture (Delayed captures method body server) new =
-  Delayed (combineRouteResults (,) captures new) method body (\ (x, v) y -> ($ v) <$> server x y)
+addCapture (Delayed captures method auth body server) new =
+  Delayed (combineRouteResults (,) captures new) method auth body (\ (x, v) y z -> ($ v) <$> server x y z)
 
 -- | Add a method check to the end of the method block.
 addMethodCheck :: Delayed a
                -> IO (RouteResult ())
                -> Delayed a
-addMethodCheck (Delayed captures method body server) new =
-  Delayed captures (combineRouteResults const method new) body server
+addMethodCheck (Delayed captures method auth body server) new =
+  Delayed captures (combineRouteResults const method new) auth body server
+
+-- | helper type family to capture server handled values for various policies
+type family AuthDelayedReturn (mP :: AuthPolicy) mE (uP :: AuthPolicy) uE usr :: * where
```

Comment:
It probably looks scarier than it is. It just captures the desired return values of the four different scenarios for auth policy:

- `'Strict x 'Strict`: handler gets a `usr`
- `'Strict x 'Lax`: handler gets a `Either e usr` where `e` is errors encountered when authenticating the user.
- `'Lax x 'Strict`: handler gets a `Either e usr` where `e` is errors encountered when extracting data from a request.
- `'Lax x 'Lax`: handler gets `Either (Either e1 e2) usr` where `e1` and `e2` are errors encountered extracting data from a request or authenticating the user.